### PR TITLE
WIP: Add the validation webhook structure

### PR DESCRIPTION
**Pull request by houshengbo**
Fixes #

## Proposed Changes

*
*
*

**Release Note**

```release-note

```

### CODE REVIEW

File: pkg/apis/operator/v1alpha1/knativeeventing_defaults.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
+/*
+Copyright 2019 The Knative Authors
```
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/knativeeventing_defaults.go:2]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: pkg/apis/operator/v1alpha1/knativeeventing_validation.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
+/*
+Copyright 2019 The Knative Authors
```
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/knativeeventing_validation.go:2]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: pkg/apis/operator/v1alpha1/knativeserving_defaults.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
+/*
+Copyright 2019 The Knative Authors
```
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/knativeserving_defaults.go:2]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: pkg/apis/operator/v1alpha1/knativeserving_validation.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
+/*
+Copyright 2019 The Knative Authors
```
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/knativeserving_validation.go:2]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: cmd/webhook/main.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	"knative.dev/pkg/injection/sharedmain"
+	"knative.dev/pkg/signals"
+	"knative.dev/pkg/webhook"
+	"knative.dev/pkg/webhook/resourcesemantics"
+	"knative.dev/pkg/webhook/resourcesemantics/validation"
+)
+
+var types = map[schema.GroupVersionKind]resourcesemantics.GenericCRD{
+	v1alpha1.SchemeGroupVersion.WithKind("KnativeEventing"):      &v1alpha1.KnativeEventing{},
+	v1alpha1.SchemeGroupVersion.WithKind("KnativeServing"): &v1alpha1.KnativeServing{},
```
Comment (mattmoor) [inline cmd/webhook/main.go:34]: Format Go code:
```suggestion
	v1alpha1.SchemeGroupVersion.WithKind("KnativeEventing"): &v1alpha1.KnativeEventing{},
	v1alpha1.SchemeGroupVersion.WithKind("KnativeServing"):  &v1alpha1.KnativeServing{},
```

File: pkg/apis/operator/v1alpha1/common_defaults.go [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+package v1alpha1
+
+import (
+	"context"
+)
+
+// SetDefaults implements apis.Defaultable
+func (cs *CommonSpec) SetDefaults(ctx context.Context) {
+	// TODO: leave it empty for further implementation
```
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/common_defaults.go:25]: Format Go code:
```suggestion
	// TODO: leave it empty for further implementation
}
```

File: pkg/apis/operator/v1alpha1/common_validation.go [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
```
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/common_validation.go:20]: Format Go code:
```suggestion
	"context"

```

File: pkg/apis/operator/v1alpha1/common_validation.go [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+	"knative.dev/pkg/apis"
+)
+
+// Validate implements apis.Validatable
+func (cs *CommonSpec) Validate(ctx context.Context) *apis.FieldError {
+	// TODO: leave it empty for future implementation
+	return nil
```
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/common_validation.go:27]: Format Go code:
```suggestion
	return nil
}
```

File: pkg/apis/operator/v1alpha1/knativeeventing_validation.go [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
```
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/knativeeventing_validation.go:20]: Format Go code:
```suggestion
	"context"

```

File: pkg/apis/operator/v1alpha1/knativeserving_types.go [hunk 10]
```diff
@@ -18,12 +18,15 @@ package v1alpha1
 
 import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/apis"
 	duckv1 "knative.dev/pkg/apis/duck/v1"
 )
 
 var (
 	_ KComponent     = (*KnativeServing)(nil)
 	_ KComponentSpec = (*KnativeServingSpec)(nil)
```
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/knativeserving_types.go:27]: Format Go code:
```suggestion
	_ KComponent       = (*KnativeServing)(nil)
	_ KComponentSpec   = (*KnativeServingSpec)(nil)
```

File: pkg/apis/operator/v1alpha1/knativeserving_validation.go [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
```
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/knativeserving_validation.go:20]: Format Go code:
```suggestion
	"context"

```

File: config/400-role.yaml [hunk 12]
```diff
@@ -371,4 +371,65 @@ rules:
...
+      - "mutatingwebhookconfigurations"
+      - "validatingwebhookconfigurations"
+    verbs: &everything
+      - "get"
+      - "list"
+      - "create"
+      - "update"
+      - "delete"
+      - "patch"
+      - "watch"
```
Comment (mattmoor) [inline config/400-role.yaml:435]: Add trailing newline:
```suggestion
      - "watch"

```

File: pkg/apis/operator/v1alpha1/common_defaults.go [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+}
```
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/common_defaults.go:26]: Add trailing newline:
```suggestion
}

```

File: pkg/apis/operator/v1alpha1/common_validation.go [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+}
```
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/common_validation.go:28]: Add trailing newline:
```suggestion
}

```

File: pkg/apis/operator/v1alpha1/knativeeventing_defaults.go (hunk 1 above)
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/knativeeventing_defaults.go:2]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: pkg/apis/operator/v1alpha1/knativeeventing_validation.go [hunk 15]
```diff
@@ -0,0 +1,44 @@
+/*
+Copyright 2019 The Knative Authors
```
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/knativeeventing_validation.go:2]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: pkg/apis/operator/v1alpha1/knativeserving_defaults.go (hunk 3 above)
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/knativeserving_defaults.go:2]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: pkg/apis/operator/v1alpha1/knativeserving_validation.go (hunk 4 above)
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/knativeserving_validation.go:2]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: cmd/webhook/main.go [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package main
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
```
Comment (mattmoor) [inline cmd/webhook/main.go:21]: Format Go code:
```suggestion
	"k8s.io/apimachinery/pkg/runtime/schema"
	"knative.dev/operator/pkg/apis/operator/v1alpha1"
	"knative.dev/operator/pkg/logconfig"
```

File: cmd/webhook/main.go [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+	"knative.dev/pkg/configmap"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/injection/sharedmain"
+	"knative.dev/pkg/signals"
+	"knative.dev/pkg/webhook"
+	"knative.dev/pkg/webhook/certificates"
+	"knative.dev/pkg/webhook/resourcesemantics"
+	"knative.dev/pkg/webhook/resourcesemantics/validation"
+	"knative.dev/operator/pkg/logconfig"
+	"knative.dev/operator/pkg/apis/operator/v1alpha1"
```
Comment (mattmoor) [inline cmd/webhook/main.go:31]: Format Go code:
```suggestion

```

File: cmd/webhook/main.go [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+)
+
+var types = map[schema.GroupVersionKind]resourcesemantics.GenericCRD{
+	v1alpha1.SchemeGroupVersion.WithKind("KnativeEventing"):      &v1alpha1.KnativeEventing{},
+	v1alpha1.SchemeGroupVersion.WithKind("KnativeServing"): &v1alpha1.KnativeServing{},
```
Comment (mattmoor) [inline cmd/webhook/main.go:36]: Format Go code:
```suggestion
	v1alpha1.SchemeGroupVersion.WithKind("KnativeEventing"): &v1alpha1.KnativeEventing{},
	v1alpha1.SchemeGroupVersion.WithKind("KnativeServing"):  &v1alpha1.KnativeServing{},
```

File: pkg/apis/operator/v1alpha1/common_defaults.go (hunk 6 above)
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/common_defaults.go:25]: Format Go code:
```suggestion
	// TODO: leave it empty for further implementation
}
```

File: pkg/apis/operator/v1alpha1/common_validation.go (hunk 7 above)
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/common_validation.go:20]: Format Go code:
```suggestion
	"context"

```

File: pkg/apis/operator/v1alpha1/common_validation.go (hunk 8 above)
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/common_validation.go:27]: Format Go code:
```suggestion
	return nil
}
```

File: pkg/apis/operator/v1alpha1/knativeeventing_validation.go [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
```
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/knativeeventing_validation.go:20]: Format Go code:
```suggestion
	"context"

```

File: pkg/apis/operator/v1alpha1/knativeserving_types.go (hunk 10 above)
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/knativeserving_types.go:27]: Format Go code:
```suggestion
	_ KComponent       = (*KnativeServing)(nil)
	_ KComponentSpec   = (*KnativeServingSpec)(nil)
```

File: pkg/apis/operator/v1alpha1/knativeserving_validation.go (hunk 11 above)
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/knativeserving_validation.go:20]: Format Go code:
```suggestion
	"context"

```

File: config/400-role.yaml (hunk 12 above)
Comment (mattmoor) [inline config/400-role.yaml:435]: Add trailing newline:
```suggestion
      - "watch"

```

File: pkg/apis/operator/v1alpha1/common_defaults.go (hunk 13 above)
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/common_defaults.go:26]: Add trailing newline:
```suggestion
}

```

File: pkg/apis/operator/v1alpha1/common_validation.go (hunk 14 above)
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/common_validation.go:28]: Add trailing newline:
```suggestion
}

```

File: cmd/webhook/main.go [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package main
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
```
Comment (mattmoor) [inline cmd/webhook/main.go:21]: Format Go code:
```suggestion
	"k8s.io/apimachinery/pkg/runtime/schema"
	"knative.dev/operator/pkg/apis/operator/v1alpha1"
	"knative.dev/operator/pkg/logconfig"
```

File: cmd/webhook/main.go [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+	"knative.dev/pkg/configmap"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/injection/sharedmain"
+	"knative.dev/pkg/signals"
+	"knative.dev/pkg/webhook"
+	"knative.dev/pkg/webhook/certificates"
+	"knative.dev/pkg/webhook/resourcesemantics"
+	"knative.dev/pkg/webhook/resourcesemantics/validation"
+	"knative.dev/operator/pkg/logconfig"
+	"knative.dev/operator/pkg/apis/operator/v1alpha1"
```
Comment (mattmoor) [inline cmd/webhook/main.go:31]: Format Go code:
```suggestion

```

File: cmd/webhook/main.go [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+)
+
+var types = map[schema.GroupVersionKind]resourcesemantics.GenericCRD{
+	v1alpha1.SchemeGroupVersion.WithKind("KnativeEventing"):      &v1alpha1.KnativeEventing{},
+	v1alpha1.SchemeGroupVersion.WithKind("KnativeServing"): &v1alpha1.KnativeServing{},
```
Comment (mattmoor) [inline cmd/webhook/main.go:36]: Format Go code:
```suggestion
	v1alpha1.SchemeGroupVersion.WithKind("KnativeEventing"): &v1alpha1.KnativeEventing{},
	v1alpha1.SchemeGroupVersion.WithKind("KnativeServing"):  &v1alpha1.KnativeServing{},
```

File: pkg/apis/operator/v1alpha1/common_defaults.go (hunk 6 above)
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/common_defaults.go:25]: Format Go code:
```suggestion
	// TODO: leave it empty for further implementation
}
```

File: pkg/apis/operator/v1alpha1/common_validation.go (hunk 7 above)
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/common_validation.go:20]: Format Go code:
```suggestion
	"context"

```

File: pkg/apis/operator/v1alpha1/common_validation.go (hunk 8 above)
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/common_validation.go:27]: Format Go code:
```suggestion
	return nil
}
```

File: pkg/apis/operator/v1alpha1/knativeeventing_validation.go [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
```
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/knativeeventing_validation.go:20]: Format Go code:
```suggestion
	"context"

```

File: pkg/apis/operator/v1alpha1/knativeserving_types.go (hunk 10 above)
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/knativeserving_types.go:27]: Format Go code:
```suggestion
	_ KComponent       = (*KnativeServing)(nil)
	_ KComponentSpec   = (*KnativeServingSpec)(nil)
```

File: pkg/apis/operator/v1alpha1/knativeserving_validation.go [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
```
Comment (mattmoor) [inline pkg/apis/operator/v1alpha1/knativeserving_validation.go:20]: Format Go code:
```suggestion
	"context"

```

**Comment by jcrossley3** [comment]
Can you add some details in the description, e.g. links to issues and/or docs, that describe what problem this is solving? Thanks!

**Pull request closed without merging**